test/e2e/storage/vsphere/vsphere_statefulsets.go

```diff
@@ -83,6 +79,7 @@ var _ = utils.SIGDescribe("vsphere statefulset [Feature:vsphere]", func() {
 		ginkgo.By("Creating statefulset")
 
 		statefulset := e2estatefulset.CreateStatefulSet(client, manifestPath, namespace)
+		defer e2estatefulset.DeleteAllStatefulSets(client, namespace)
```

Comment:
https://github.com/kubernetes/kubernetes/issues/92785

> Issue-2 - After upgrading apiVersion from apps/v1. Test executed successfully but reported failure in the after each cleanup stage.
> Jul  3 20:34:46.049: FAIL: Failed waiting for stateful set status.replicas updated to 0: statefulsets.apps "web" not found
> ...
> This is happening because of the incorrect usage of WaitForStatusReplicas after calling c.AppsV1().StatefulSets(ss.Namespace).Delete.

trying to understand more here and why this change matters.

are you suggesting that DeleteAllStatefulSets() gets called before WaitForStatusReplicas() if the DeleteAllStatefulSets() call is inside AfterEach(...)? are these tests run in parallel when you saw the problem?

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/vsphere/vsphere_statefulsets.go#L131

cc @SandeepPissay
vSphere storage

cc @gnufied @pohly 
SIG Storage / e2e

Reply:
@neolit123 

> why this change matters

We want to avoid printing below error message in the test for `AfterEach` call.

This could be because of the race between namespace deletion and AfterEach execution. I am not familiar with the framework. When `c.AppsV1().StatefulSets(ss.Namespace).Delete` is called, it is not finding a stateful set.

```
[AfterEach] [sig-storage] vsphere statefulset [Feature:vsphere]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_statefulsets.go:69
Jul  3 20:34:25.949: INFO: Deleting all statefulset in namespace: vsphere-statefulset-4079
Jul  3 20:34:25.964: INFO: Scaling statefulset web to 0
Jul  3 20:34:46.036: INFO: Waiting for statefulset status.replicas updated to 0
Jul  3 20:34:46.049: FAIL: Failed waiting for stateful set status.replicas updated to 0: statefulsets.apps "web" not found

Full Stack Trace
k8s.io/kubernetes/test/e2e/framework/statefulset.WaitForStatusReplicas(0x5e49f80, 0xc00083c6e0, 0xc000cf6f00, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/statefulset/wait.go:145 +0x22e
k8s.io/kubernetes/test/e2e/framework/statefulset.DeleteAllStatefulSets(0x5e49f80, 0xc00083c6e0, 0xc0035a80a0, 0x18)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/statefulset/rest.go:86 +0x3a9
k8s.io/kubernetes/test/e2e/storage/vsphere.glob..func5.2()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/vsphere/vsphere_statefulsets.go:71 +0xe0
k8s.io/kubernetes/test/e2e.RunE2ETests(0xc000124ea0)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:130 +0x337
k8s.io/kubernetes/test/e2e.TestE2E(0xc000124ea0)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:145 +0x2b
testing.tRunner(0xc000124ea0, 0x55e7630)
	/usr/local/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1042 +0x357

• Failure in Spec Teardown (AfterEach) [146.105 seconds]
```